filebeat/harvester/reader/multiline.go

```diff
@@ -149,86 +152,103 @@ func (mlr *Multiline) readNext() (Message, error) {
 
 				// return collected multiline event and
 				// empty buffer for new multiline event
-				p := mlr.pushLine()
-				mlr.reset()
-				return p, nil
+				message := mlr.flush()
+				mlr.resetState()
+				return message, nil
 			}
 
 			// handle error without any bytes returned from reader
-			if p.Bytes == 0 {
+			if message.Bytes == 0 {
 				// no lines buffered -> return error
 				if mlr.numLines == 0 {
 					return Message{}, err
 				}
 
 				// lines buffered, return multiline and error on next read
-				p := mlr.pushLine()
+				m := mlr.flush()
 				mlr.err = err
-				mlr.state = (*Multiline).readFailed
-				return p, nil
+				mlr.setState((*Multiline).readFailed)
+				return m, nil
 			}
 
 			// handle error with some content being returned by reader and
 			// line matching multiline criteria or no multiline started yet
-			if mlr.readBytes == 0 || mlr.pred(mlr.last, p.Content) {
-				mlr.addLine(p)
+			if mlr.readBytes == 0 || mlr.pred(mlr.last, message.Content) {
+				mlr.addLine(message)
 
 				// return multiline and error on next read
-				l := mlr.pushLine()
+				m := mlr.flush()
 				mlr.err = err
-				mlr.state = (*Multiline).readFailed
-				return l, nil
+				mlr.setState((*Multiline).readFailed)
+				return m, nil
 			}
 
 			// no match, return current multline and retry with current line on next
 			// call to readNext awaiting the error being reproduced (or resolved)
 			// in next call to Next
-			l := mlr.startNewLine(p)
-			return l, nil
+			m := mlr.flush()
+			mlr.init(message)
+			return m, nil
 		}
 
 		// if predicate does not match current multiline -> return multiline event
-		if mlr.readBytes > 0 && !mlr.pred(mlr.last, p.Content) {
-			l := mlr.startNewLine(p)
-			return l, nil
+		if mlr.readBytes > 0 && !mlr.pred(mlr.last, message.Content) {
+			m := mlr.flush()
+			mlr.init(message)
+			return m, nil
 		}
 
 		// add line to current multiline event
-		mlr.addLine(p)
+		mlr.addLine(message)
 	}
 }
 
+// readFailed returns empty message and error and resets line reader
 func (mlr *Multiline) readFailed() (Message, error) {
-	// return error and reset line reader
 	err := mlr.err
 	mlr.err = nil
-	mlr.reset()
+	mlr.resetState()
 	return Message{}, err
 }
 
-func (mlr *Multiline) startNewLine(l Message) Message {
-	retLine := mlr.pushLine()
-	mlr.addLine(l)
-	mlr.ts = l.Ts
-	mlr.fields = l.Fields
-	return retLine
+// inits the reader with the given message. It is recommend to either
+// run reset or flush before.
+func (mlr *Multiline) init(m Message) {
+	mlr.addLine(m)
+	// Timestamp of first message is taken as overall timestamp
+	mlr.ts = m.Ts
+	mlr.fields = m.Fields
 }
 
-func (mlr *Multiline) pushLine() Message {
-	content := mlr.content
-	sz := mlr.readBytes
-	fields := mlr.fields
+// flush writes the existing content into the returned message and resets all reader variables.
+func (mlr *Multiline) flush() Message {
```

Comment:
`finalize` and `finish` also both sound like they don't return something. I like `flush()` because it somehow implies the reset afterwards, but I agree that it is not obvious where it flushes to ... Alternatives: `fetch`, `consume` ...?